File: typescript/src/services/TokenManager.ts
import { ICredentialsRepository } from '../interfaces/ICredentialsRepository'
import { ITokenManager } from '../interfaces/ITokenManager'

/**
 * Gerenciador de tokens OAuth com cache e refresh automático.
 */
export class TokenManager implements ITokenManager {
  private tokenCache: Record<string, Record<string, any>> = {}

  constructor(private readonly credentialsRepository: ICredentialsRepository) {}

  /**
   * Obtém token de acesso válido.
   * @param id Identificador
   * @returns Promise com token válido
   */
  async getAccessToken(id: string): Promise<string> {
    try {
      // Verifica cache primeiro
      const cached = this.tokenCache[id]
      if (cached && this.isTokenValid(cached.validade)) {
        return cached.access_token
      }

      // Se não tem no cache ou está expirado, busca do repositório
      const stored = await this.getTokenInRepository(id)

      // Se tem token armazenado e é válido, atualiza cache e retorna
      if (stored && this.isTokenValid(stored.validade)) {
        this.tokenCache[id] = stored
        return stored.access_token
      }

      // Se tem refresh token, tenta atualizar
      if (stored?.refresh_token) {
        const refreshed = await this.refreshToken(id, stored)
        return refreshed.access_token
      }

      // Se nada funcionar, obtém novo token
      const newToken = await this._obtainNewToken(id)
      await this.credentialsRepository.saveToken(id, newToken)
      this.tokenCache[id] = newToken
      return newToken.access_token
    } catch (error) {
      console.error(`Erro ao obter access token para ${id}:`, error)
      throw error
    }
  }

  /**
   * Obtém token do repositório.
   * @param id Identificador
   * @returns Promise com o token
   */
  async getTokenInRepository(id: string): Promise<Record<string, any>> {
    try {
      const credentials = await this.credentialsRepository.getCredentials(id)
      return credentials
    } catch (error) {
      console.error(`Erro ao obter token do repositório para ${id}:`, error)
      throw error
    }
  }

  /**
   * Verifica se o token é válido.
   * @param validity Data de validade
   * @returns true se válido
   */
  isTokenValid(validity: string): boolean {
    try {
      if (!validity) return false
      const validUntil = new Date(validity).getTime()
      const now = new Date().getTime()
      return validUntil > now
    } catch {
      return false
    }
  }

  /**
   * Atualiza token usando refresh token.
   * @param id Identificador
   * @param cred Credenciais atuais
   * @returns Promise com novo token
   */
  async refreshToken(
    id: string,
    cred: Record<string, any>
  ): Promise<Record<string, any>> {
    try {
      // Implementar lógica de refresh específica da sua API
      throw new Error('Método refreshToken não implementado')
    } catch (error) {
      console.error(`Erro ao atualizar token para ${id}:`, error)
      throw error
    }
  }

  /**
   * Força atualização do token.
   * @param id Identificador
   * @returns Promise com novo token
   */
  async forceRefreshingToken(id: string): Promise<string> {
    try {
      const stored = await this.getTokenInRepository(id)
      const refreshed = await this.refreshToken(id, stored)
      await this.credentialsRepository.saveToken(id, refreshed)
      this.tokenCache[id] = refreshed
      return refreshed.access_token
    } catch (error) {
      console.error(`Erro ao forçar atualização do token para ${id}:`, error)
      throw error
    }
  }

  /**
   * Obtém novo token via OAuth.
   * @param id Identificador
   * @returns Promise com novo token
   */
  async _obtainNewToken(id: string): Promise<Record<string, any>> {
    // Implementar lógica de obtenção de novo token específica da sua API
    throw new Error('Método _obtainNewToken não implementado')
  }
}

File: python/src/services/token_manager.py
"""
Gerenciador de tokens implementando ITokenManager.

Gerencia tokens OAuth com cache e refresh automático.
"""

from datetime import datetime
from typing import Any, Dict

import pandas as pd

from ..interfaces.credentials_repository_interface import ICredentialsRepository
from ..interfaces.token_manager_interface import ITokenManager
from ..utils.log import log


class TokenManager(ITokenManager):
    """Gerenciador de tokens OAuth com cache e refresh automático."""

    def __init__(
        self,
        credentials_repository: ICredentialsRepository,
    ):
        """
        Inicializa o gerenciador de tokens.

        Args:
            credentials_repository: Repositório de credenciais
        """
        self._credentials_repository = credentials_repository
        self._token_cache: Dict[str, Dict[str, Any]] = {}

    def get_access_token(self, id: str) -> str:
        """
        Obtém token de acesso válido para a loja.

        Args:
            id: Identificador da loja

        Returns:
            Token de acesso válido
        """

        try:
            # Verifica cache primeiro, Se não estiver no cache, busca no repositório
            if id in self._token_cache:
                cred = self._token_cache[id]
            else:
                cred = self._credentials_repository.get_credentials(id)

            if not self.is_token_valid(cred.get("validade", "")):
                log.info(f"Token inválido para {id}, atualizando...")
                cred = self.refresh_token(id, cred)
                self._credentials_repository.save_token(id, cred)

            self._token_cache[id] = cred

            return self._token_cache[id]["access_token"]

        except Exception as e:
            log.error(f"Erro ao obter token para {id}: {e}")
            raise

    def is_token_invalid(self, validade: str) -> bool:
        """
        Verifica se o token da loja é válido.

        Args:
            store: Identificador da loja

        Returns:
            True se o token for válido
        """
        try:
            if not validade or validade == "":
                return False

            expiration = pd.to_datetime(validade)
            return datetime.now() > expiration

        except Exception:
            return False

    def refresh_token(self, id: str, cred: Dict[str, Any]) -> Dict[str, Any]:
        """
        Atualiza token usando refresh_token.

        Args:
            store: Identificador da loja

        Returns:
            Novo token de acesso
        """

        try:
            return {"access_token": "", "refresh_token": "", "validade": ""}

        except Exception as e:
            log.error(f"Erro ao fazer refresh do token para {id}: {e}")
            return self._obtain_new_token(id)

    def force_refreshing_token(self, id: str) -> str:
        """
        Força a atualização do token de acesso.

        Args:
            id: Identificador da loja

        Returns:
            Novo token de acesso
        """
        try:
            if id in self._token_cache:
                cred = self._token_cache[id]
            else:
                cred = self.get_token_in_repository(id)
                cred = {
                    "access_token": cred["access_token"],
                    "refresh_token": cred["refresh_token"],
                    "validade": cred["validade"],
                }

            cred = self.refresh_token(id, cred)
            self._credentials_repository.save_token(id, cred)
            return cred["access_token"]

        except Exception as e:
            log.error(f"Erro ao forçar refresh do token para {id}: {e}")
            raise

    def _obtain_new_token(self, id: str) -> Dict[str, Any]:
        """
        Obtém novo token através do fluxo OAuth.

        Args:
            store: Identificador da loja

        Returns:
            Novo token de acesso
        """
        return {"access_token": "", "refresh_token": "", "validade": ""}
